### PR TITLE
WIP: fromFHIR implementations

### PR DESCRIPTION
**Pull request by dehall**
Work in progress on developing fromFHIR that mirrors the latest toFHIR generation. Submitting this PR before totally complete to get some early feedback. Remaining to-do items listed below.
Also note this PR is based off the `to_fhir_slicing` branch.

The goal here is that for any generated class, fromFHIR is the "inverse" of toFHIR, so that `fromFHIR(toFHIR(shr)) == shr` and `toFHIR(fromFHIR(fhir)) == fhir`. The fromFHIR function is generated by iterating over the fields of the FHIR profile and building assignment statements as necessary. Extensions are handled with a lookup based on URL.  One tricky point is that because of SHR class inheritance and Options, it's not necessarily possible to know what the type of an field is going to be beforehand. So in general we can't just do `inst.field = BaseClass.fromFHIR(fhir['field'])` .  Instead I added a new `createInstanceFromFHIR` function that takes a suggested type name and FHIR, and will lookup the profile from the FHIR to pick the SHR type or fallback to the given type. So now most of the generated assignments look like `inst.field = createInstanceFromFHIR('shr.spec.BaseClass', fhir['field']);` This makes its way through the factory classes just like the current `createInstance` does.

For some examples, here are Person.js and Patient.js generated off the master branch of shr_spec:

[Patient.js.txt](https://github.com/standardhealth/shr-es6-export/files/2487966/Patient.js.txt)
[Person.js.txt](https://github.com/standardhealth/shr-es6-export/files/2487967/Person.js.txt)

Still to do:
- [ ] Add and improve test cases
- [ ] Address any issues found from test cases
- [ ] Address a bug with nested field cardinality which also exists in toFHIR (ex. the FHIR Patient profile maps Patient.address to SHR Patient.person.address and both are 0..* cardinality, but as of now the generated toFHIR and fromFHIR will only map one element of the array... tough to explain in one sentence but I can elaborate if desired)
- [ ] Address lint issue in generated code, where `createInstanceFromFHIR` is always imported but not always used.
- [ ] Review any `TODO` in the code
- [ ] General cleanup including style and comments/documentation

### CODE REVIEW

**Comment by dehall** [comment]
Update: this still needs work (potentially a lot) to handle slicing correctly. Example of where it's at now (i.e., broken) based on the BloodPressureEntry test fixture:

```
  toFHIR(asExtension=false) {
    let inst = {};
    inst['resourceType'] = 'Observation';
    if (this.systolicPressure != null) {
      inst['component'] = inst ['component'] || [];
      inst['component'].push(typeof this.systolicPressure.toFHIR === 'function' ? this.systolicPressure.toFHIR() : this.systolicPressure);
    }
    if (this.diastolicPressure != null) {
      inst['component'] = inst ['component'] || [];
      inst['component'].push(typeof this.diastolicPressure.toFHIR === 'function' ? this.diastolicPressure.toFHIR() : this.diastolicPressure);
    }
    return inst;
  }

  static fromFHIR(fhir, asExtension=false) {
    const inst = new BloodPressureEntry();
    if (fhir['component'] != null) {
      inst.systolicPressure = createInstanceFromFHIR('shr.slicing.SystolicPressure', fhir['component'][0]);
    }
    if (fhir['component'] != null) {
      inst.diastolicPressure = createInstanceFromFHIR('shr.slicing.DiastolicPressure', fhir['component'][0]);
    }
    return inst;
  }

```

**Comment by dehall** [comment]
Update: that last commit fixed one field cardinality issue but broke something else. Patient.person.address works correctly now, but BarAEntry in the test fixtures here is now incorrect. Looking now into how to get both working at the same time

File: lib/generateClass.js [hunk 1]
```diff
@@ -284,14 +307,222 @@ function writeToJson(def, cw) {
...
+          const url = element.type[0].profile;
+          const varName = `match_${i}`; // ensure a unique variable name here
+          cw.ln(`const ${varName} = fhir['extension'].find(e => e.url == '${url}');`);
+          cw.bl(`if (${varName} != null)`, () => {
+            cw.ln(`inst.${methodName} = createInstanceFromFHIR('${instance}', ${varName}, true);`); // asExtension = true here, false(default value) everywhere else
+          });
+        }
+      });
+    });
+  }
```
Comment (cmoesel) [inline lib/generateClass.js]: I'll be honest with you...  I'm having trouble wrapping my head around the extension stuff above.  I'd recommend heavily testing it, and if it works, I'll trust that it's right! ;-)

File: lib/generateClass.js [hunk 2]
```diff
@@ -284,14 +307,222 @@ function writeToJson(def, cw) {
...
+  if (!fhirProfile && def.value) {
+    // If not a profile and can be resolved directly to a value, set it as the "value" field
+    cw.bl(`if (!asExtension && fhir != null)`, () => {
+
+      if (def.value.identifier != null) {
+        if (def.value.identifier.isPrimitive) {
+          // it's a primitive value means we can set it directly
+          cw.ln(`inst.value = fhir;`);
+        } else {
+          const shrType = def.value.identifier.fqn;
```
Comment (cmoesel) [inline lib/generateClass.js]: Again, I think you want `effectiveIdentifier` here?

File: lib/generateClass.js [hunk 3]
```diff
@@ -350,12 +581,25 @@ function writeToFhir(def, fhir, cw) {
...
+            // for nested element mappings, such as "Patient.person.address", we have to look up Patient.person and then Person.address
+            for(const pathElement of fieldMapPath) {
+              field = currDef.fields.find(f => f.identifier && f.identifier.equals(fhirMappingToIdentifier(pathElement)));
+
+              if (!field) {
+                break;
+              }
+
+              currDef = specs.dataElements.byNamespace(field.identifier.namespace).find(e =>  e.identifier.equals(field.identifier));
+            }
```
Comment (cmoesel) [inline lib/generateClass.js]: This block (lines 584 - 596) looks very similar to code I commented on up above.  If so, consider breaking it out into a re-usable function -- especially since I think I suggested changes to be made in the logic.

File: lib/generateClass.js [hunk 4]
```diff
@@ -350,12 +581,25 @@ function writeToFhir(def, fhir, cw) {
...
+
+            if(field){
```
Comment (cmoesel) [inline lib/generateClass.js]: If `field` doesn't exist, is that an ok (expected) condition?  Or is it something we should log an error about?

File: lib/generateClass.js [hunk 5]
```diff
@@ -549,6 +795,182 @@ function generateFHIRAssignment(cardIsList, baseCardIsList, constraintsLength, f
...
+  if (!(cardIsList || baseCardIsList) && constraintsLength > 1) {
+    const pathArrayClone = fhirElementPath.split('.');
+    // loop over the pathArrayClone to figure out where th nearest 0..*-cardinality element is
+    while (pathArrayClone.length > 0) {
+      pathArrayClone.pop();
+      const searchString = pathArrayClone.join('.');
+      const candidateSnapshotElement = fhirProfile.snapshot.element.find(e => { return e.path === searchString; });
+      // If we find a snapshot element that matches the search string, and which has a 'max' (cardinality) of *
+      // That's our nearest array element (into which we're going to put our elements)
+      if (candidateSnapshotElement && candidateSnapshotElement.max == '*') {
```
Comment (cmoesel) [inline lib/generateClass.js]: I think you also want to allow for cases where `max` is a number > 1 (but note it will be a stringified number).  For example, what if `max` is `"5"`?

File: lib/generateClass.js [hunk 6]
```diff
@@ -549,6 +795,182 @@ function generateFHIRAssignment(cardIsList, baseCardIsList, constraintsLength, f
...
+          } else {
+            const shrMapping = shrElementMapping[shrElementMapping.length - 1];
+            const shrType = fhirMappingToIdentifier(shrMapping).fqn;
+            cw.ln(`inst.${shrElementPath} = createInstanceFromFHIR('${shrType}', fhir${fhirPathString});`);
+          }
+        }
+      }
+    }
+  });
+}
```
Comment (cmoesel) [inline lib/generateClass.js]: My brain is turning to mush and code generation code is really hard to follow because you need to concurrently be building and remembering the generated code in your head while also reviewing the code that does the generation.  In a way, it's a bit like a meta-code review.

So... I only looked at the code above briefly once I realized my brain was not processing it well anyway.  Instead I'll probably just try it and review the output.

File: test/es6FromFHIRJSONTest.js [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+describe('#FromFHIR', () => {
+  describe('#PatientDirectMapEntry()', () => {
+    const PatientDirectMapEntry = importResult('shr/fhir/PatientDirectMapEntry');
+    it('should deserialize a FHIR JSON instance', () => {
+      const json = context.getFHIR('PatientDirectMapEntry');
+      const entry = PatientDirectMapEntry.fromFHIR(json);
+      expect(entry).instanceOf(PatientDirectMapEntry);
+
+      const roundTrip = entry.toFHIR();
+      expect(roundTrip).to.eql(json);
```
Comment (cmoesel) [inline test/es6FromFHIRJSONTest.js]: So a round-trip test is a good idea, but if there are bugs in `toFHIR` (and I suspect there are), then round-tripping only ensures that you have corresponding bugs in `fromFHIR`!  I'd like to see at least _some_ tests that actually inspect the result to ensure that all of the data is where it should be in the class instance.

File: lib/generateClass.js [hunk 8]
```diff
@@ -284,14 +307,222 @@ function writeToJson(def, cw) {
...
+    // if the fhir profile ends and we still have "open" slices, finalize them here
+    while (sliceStack.length > 0) {
+      const sliceGroup = sliceStack.shift();
+      sliceMap[sliceGroup.path] = sliceGroup.elements;
+    }
+
+    for (let element of fhirProfile.snapshot.element) {
+      const mapping = element.mapping && element.mapping.find(m => m['identity'] === 'shr');
+      let baseIsList = element.base && element.base.max == '*';
+      baseIsList = baseIsList || (element.mapping && element.mapping.filter(elem => elem.identity == 'shr').length > 1);
```
[the remark this answers is not in the document]
Comment (dehall) [reply lib/generateClass.js]: This is based on the toFHIR logic which does the same thing. I haven't seen any examples of multiple mappings so far, but I'll look into it, and if necessary will update toFHIR as well.

File: lib/generateClass.js [hunk 9]
```diff
@@ -284,14 +307,222 @@ function writeToJson(def, cw) {
...
+              }
+
+              currDef = specs.dataElements.byNamespace(field.identifier.namespace).find(e =>  e.identifier.equals(field.identifier));
+            }
+
+            if(field){
+              // console.log(field.identifier.name);
+              if (!alreadyMappedElements.has(classMethodChain)) {
+                const constraintsLength = field.constraintsFilter.includesType.constraints.length;
+                const slicing = sliceMap[element.path] && sliceMap[element.path][element.sliceName];
```
[the remark this answers is not in the document]
Comment (dehall) [reply lib/generateClass.js]: Are you sure the `&&` converts it to true/false? I'm intending to use this as a "guard operator" where it will only try to access `sliceMap[element.path][element.sliceName]` if `sliceMap[element.path]` is not null. I figured the one line version is shorter than the traditional null check.

[the remark this answers is not in the document]
Comment (dehall) [reply lib/generateClass.js]: But I definitely have to look into whether `element.sliceName` will always be present here. From the examples I saw the SHR mapping was always on the same element as the sliceName, but maybe that won't necessarily be the case.

File: lib/generateClass.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (cmoesel) [reply lib/generateClass.js]: OK.  Thanks.  Like I said, may totally be legit -- but the code here definitely does imply there might be more than one.

File: lib/generateClass.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (cmoesel) [reply lib/generateClass.js]: Wow.  You are 100% right, @dehall .  TIL that I've been misunderstanding `&&` in JavaScript for a _long_ time.  Thanks for setting me straight!

File: lib/generateClass.js [hunk 10]
```diff
@@ -284,14 +307,222 @@ function writeToJson(def, cw) {
...
+            let field;
+            let currDef = def;
+
+            // for nested element mappings, such as "Patient.person.address", we have to look up Patient.person and then Person.address
+            for(const pathElement of fieldMapPath) {
+              field = currDef.fields.find(f => f.identifier && f.identifier.equals(fhirMappingToIdentifier(pathElement)));
+
+              if (!field) {
+                break;
+              }
```
[the remark this answers is not in the document]
Comment (dehall) [reply lib/generateClass.js]: In researching this, I found a couple quick fixes to find the field here, specifically it could be the `value`. But there are still a lot of instances where the matching field is not found. 

My sense here is that there shouldn't be cases where the field isn't found, but I'm not yet sufficiently well-versed in the SHR/CIMPL ecosystem to resolve or explain the remaining cases. So rather than continue down that rabbit hole at the moment, I'm going to leave the `break` but add some error logging. Will follow up on this one once I've addressed the other open items.

File: lib/generateClass.js [hunk 11]
```diff
@@ -284,14 +307,222 @@ function writeToJson(def, cw) {
...
+        sliceStack.unshift(sliceGroup);
+      } else if (sliceStack[0]) {
+
+        if (element.sliceName) {
+          // starting a new slice element, only possible if already within a slice group
+
+          sliceStack[0].currentSlice = element.sliceName;
+          sliceStack[0].elements[element.sliceName] = [];
+
+        } else {
```
[the remark this answers is not in the document]
Comment (dehall) [reply lib/generateClass.js]: Pretty much all of your comments here have been automatically marked as "outdated" because of significant code refactoring, other than this one. This one refers to the section which is now in function `preprocessSlicing` (around line 826 currently). I did move that section out of the else block, so I'm going to mark this as resolved, and try and find and double check the other comments.

**Comment by dehall** [comment]
Closing this PR and will open a new one against`master`.
